src/core/structure/sheet.ts

```diff
@@ -81,6 +81,87 @@ export default class Sheet {
     return cell ? cell.value : null;
   }
 
+  moveColumn(from: number, to: number): boolean {
+    if (from === to) return false;
+
+    const colKey = this.columnPositions.get(from);
+
+    if (colKey !== undefined) {
+      const col = this.columns.get(colKey);
+      if (col === undefined)
+        throw new Error("Column not found, inconsistent state");
+      col.position = to;
+    }
+
+    //we have to update all the positions to keep it consistent
+    this.columnPositions.delete(from);
+    if (to > from) {
+      this.shiftColumns(to, ShiftDirection.backward);
+    } else {
+      this.shiftColumns(to, ShiftDirection.forward);
+    }
+    if (colKey === undefined) {
+      this.columnPositions.delete(to);
+    } else {
+      this.columnPositions.set(to, colKey);
+    }
+
+    return true;
+  }
+
+  insertColumn(position: number): boolean {
+    this.shiftColumns(position, ShiftDirection.forward);
+    return true;
+  }
+
+  deleteColumn(position: number): boolean {
+    const colKey = this.columnPositions.get(position);
+
+    if (colKey !== undefined) {
+      const col = this.columns.get(colKey);
+      for (const [rowKey] of col!.cellIndex) {
+        this.deleteCell(colKey, rowKey);
+      }
+    }
+
+    const lastColumnPosition = Math.max(...this.columnPositions.keys());
+    this.shiftColumns(lastColumnPosition, ShiftDirection.backward);
```

Comment:
This doesn't seem correct. Shouldn't we pass the deleted position to `shiftColumns`? Consider the following case:

![image](https://github.com/lightsheet-team/lightsheet/assets/38987331/18153b97-81c1-419f-9b94-58358f00deb7)
If column C is deleted, we don't want to shift anything on the left side of column C. However, `shiftColumns` currently shifts all values from max column to the left, meaning values in B are shifted to A. Here's also a test (run in `deleteColumn.ts`) illustrating the issue:
```ts
  it("Should delete the last column and not affect other columns", () => {
    sheet.setCellAt(0, 0, "");
    sheet.setCellAt(0, 1, "");
    console.log(sheet.exportData());
    sheet.deleteColumn(2);
    console.log(sheet.exportData());
    expect(sheet.getCellValueAt(0, 0)).toBe(null);
    expect(sheet.getCellValueAt(1, 0)).toBe("2x1");
    expect(sheet.getCellValueAt(2, 0)).toBe(null);
    expect(sheet.getCellValueAt(0, 1)).toBe(null);
    expect(sheet.getCellValueAt(1, 1)).toBe("2x2");
    expect(sheet.getCellValueAt(2, 1)).toBe(null);
  });

Reply:
Good catch! 
I added your test to the code and fixed it.
 
> Shouldn't we pass the deleted position to shiftColumns?

You can think of a shift operation as something pulling the values from the `position ` towards the first empty column position in a certain direction. So no, we shouldn't pass the deleted position, at least that's how I made it.

The fix is simply not shifting if we are deleting the last column because it is not needed. So I added a condition for that.

Reply:
I see now; I misunderstood the underlying issue :+1: